File: pt7/1334.find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.java
import java.util.Arrays;

/*
 * LeetCode 1334 - Find the City With the Smallest Number of Neighbors at a Threshold Distance
 */
class Solution {
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        final int inf = 1_000_000_000;
        int[][] dist = new int[n][n];

        for (int i = 0; i < n; i++) {
            Arrays.fill(dist[i], inf);
            dist[i][i] = 0;
        }

        for (int[] edge : edges) {
            int a = edge[0];
            int b = edge[1];
            int weight = edge[2];
            dist[a][b] = Math.min(dist[a][b], weight);
            dist[b][a] = Math.min(dist[b][a], weight);
        }

        for (int mid = 0; mid < n; mid++) {
            for (int start = 0; start < n; start++) {
                for (int end = 0; end < n; end++) {
                    dist[start][end] = Math.min(
                            dist[start][end],
                            dist[start][mid] + dist[mid][end]);
                }
            }
        }

        int bestCity = -1;
        int bestCount = n + 1;

        for (int city = 0; city < n; city++) {
            int reachable = 0;
            for (int other = 0; other < n; other++) {
                if (other != city && dist[city][other] <= distanceThreshold) {
                    reachable++;
                }
            }

            if (reachable <= bestCount) {
                bestCount = reachable;
                bestCity = city;
            }
        }

        return bestCity;
    }
}

/*
 * Interview explanation
 * ---------------------
 * Algorithm:
 * We need all-pairs shortest paths. With n <= 100, Floyd-Warshall is simple
 * and reliable: for every possible intermediate city, try to improve every
 * start-to-end distance.
 *
 * Java data structures:
 * `int[][] dist` stores shortest known distances. `Arrays.fill` initializes
 * each row to a large sentinel value.
 *
 * Tie-breaking:
 * The problem wants the greatest city id when counts tie. We scan from low to
 * high and update on `<=`, so later city ids replace earlier ids on ties.
 *
 * Edge cases:
 * - Disconnected pairs remain at infinity and are not counted.
 * - Multiple edges between two cities keep the smallest weight.
 * - Threshold 0 generally counts no other city.
 *
 * Complexity:
 * Time O(n^3).
 * Space O(n^2).
 */

File: Google/1334.find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.py
#
# @lc app=leetcode id=1334 lang=python3
#
# [1334] Find the City With the Smallest Number of Neighbors at a Threshold Distance
#
# https://leetcode.com/problems/find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/description/
#
# algorithms
# Medium (73.52%)
# Likes:    3686
# Dislikes: 160
# Total Accepted:    353K
# Total Submissions: 480K
# Testcase Example:  "4"
#
# There are n cities numbered from 0 to n-1. Given the array edges where
# edges[i] = [from_i, to_i, weight_i] represents a bidirectional and weighted
# edge between cities from_i and to_i, and given the integer distanceThreshold.
#
# Return the city with the smallest number of cities that are reachable through
# some path and whose distance is at most distanceThreshold, If there are
# multiple such cities, return the city with the greatest number.
#
# Notice that the distance of a path connecting cities i and j is equal to the
# sum of the edges' weights along that path.
#
# Example 1:
#
# Input: n = 4, edges = [[0,1,3],[1,2,1],[1,3,4],[2,3,1]], distanceThreshold =
# 4
# Output: 3
# Explanation: The figure above describes the graph.
# The neighboring cities at a distanceThreshold = 4 for each city are:
# City 0 -> [City 1, City 2]
# City 1 -> [City 0, City 2, City 3]
# City 2 -> [City 0, City 1, City 3]
# City 3 -> [City 1, City 2]
# Cities 0 and 3 have 2 neighboring cities at a distanceThreshold = 4, but we
# have to return city 3 since it has the greatest number.
#
# Example 2:
#
# Input: n = 5, edges = [[0,1,2],[0,4,8],[1,2,3],[1,4,2],[2,3,1],[3,4,1]],
# distanceThreshold = 2
# Output: 0
# Explanation: The figure above describes the graph.
# The neighboring cities at a distanceThreshold = 2 for each city are:
# City 0 -> [City 1]
# City 1 -> [City 0, City 4]
# City 2 -> [City 3, City 4]
# City 3 -> [City 2, City 4]
# City 4 -> [City 1, City 2, City 3]
# The city 0 has 1 neighboring city at a distanceThreshold = 2.
#
# Constraints:
#
# 2 <= n <= 100
#
# 1 <= edges.length <= n * (n - 1) / 2
#
# edges[i].length == 3
#
# 0 <= from_i < to_i < n
#
# 1 <= weight_i, distanceThreshold <= 10^4
#
# All pairs (from_i, to_i) are distinct.
#

# @lc code=start
from typing import List


class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        """
        Interview explanation:
        Undirected weighted graph. For each city count others reachable within
        threshold. Pick city with smallest count; ties -> largest city id.
        Floyd-Warshall is classic for all-pairs on small n (<=100).

        Algorithm (Floyd-Warshall):
        - dist[][] init INF/0/edge; triple loop relax; count per row; pick answer.

        Complexity: O(n^3) time, O(n^2) space.
        """
        INF = 10**9
        dist = [[INF] * n for _ in range(n)]
        for i in range(n):
            dist[i][i] = 0
        for u, v, w in edges:
            dist[u][v] = dist[v][u] = w
        for k in range(n):
            for i in range(n):
                dik = dist[i][k]
                if dik == INF:
                    continue
                for j in range(n):
                    cand = dik + dist[k][j]
                    if cand < dist[i][j]:
                        dist[i][j] = cand

        best_cnt, ans = n, 0
        for i in range(n):
            cnt = sum(dist[i][j] <= distanceThreshold for j in range(n) if j != i)
            if cnt <= best_cnt:
                best_cnt, ans = cnt, i
        return ans

    def findTheCity_dijkstra(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        """
        Interview explanation:
        Alternate: Dijkstra from each city (or Bellman-Ford). Better when sparse.

        Algorithm:
        - Build adj; for each source Dijkstra; count; pick city.

        Complexity: O(n * (n+m) log n) time, O(n+m) space.
        """
        import heapq
        from collections import defaultdict

        g = defaultdict(list)
        for u, v, w in edges:
            g[u].append((v, w))
            g[v].append((u, w))

        def count_reach(src: int) -> int:
            dist = [10**9] * n
            dist[src] = 0
            pq = [(0, src)]
            while pq:
                d, u = heapq.heappop(pq)
                if d > dist[u]:
                    continue
                for v, w in g[u]:
                    nd = d + w
                    if nd < dist[v]:
                        dist[v] = nd
                        heapq.heappush(pq, (nd, v))
            return sum(1 for i in range(n) if i != src and dist[i] <= distanceThreshold)

        best_cnt, ans = n, 0
        for i in range(n):
            cnt = count_reach(i)
            if cnt <= best_cnt:
                best_cnt, ans = cnt, i
        return ans
# @lc code=end

